vertx-sql-client/src/main/java/io/vertx/sqlclient/Query.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2017 Julien Viet
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.vertx.sqlclient;
+
+import io.vertx.codegen.annotations.GenIgnore;
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+/**
+ * A query.
+ *
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+@VertxGen
+public interface Query<T> {
+
+  /**
+   * Execute the query.
+   *
+   * @param handler the handler receiving the response
+   */
+  void execute(Handler<AsyncResult<T>> handler);
+
+  /**
+   * Like {@link #execute(Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  Future<T> execute();
+
+  /**
+   * Execute the query.
+   *
+   * @param handler the handler receiving the response
+   */
+  void execute(Tuple tuple, Handler<AsyncResult<T>> handler);
+
+  /**
+   * Like {@link #execute(Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  Future<T> execute(Tuple tuple);
+
+  /**
+   * Execute the query.
+   *
+   * @param batch the batch of tuples
+   * @param handler the handler receiving the response
+   */
+  void batch(List<Tuple> batch, Handler<AsyncResult<T>> handler);
+
+  /**
+   * Like {@link #batch(List, Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  Future<T> batch(List<Tuple> batch);
```

Comment:
I think we need to have a separate `PreparedQuery` interface, because the methods that accept a `Tuple` argument are only for prepared queries.
Currently both `createQuery()` and `createPreparedQuery()` return type `Query`. As @BillyYccc pointed out a user could try to send in params to a non-prepared query, which is invalid:
```java
// Error: cannot set params on a non-prepared query
client.createQuery(SQL).execute(Tuple.of("foo"));
```

Reply:
I think it's fine like that and we do not need to over engineer this

Reply:
My point is that the `PreparedQuery` would be very similar to `Query` with only the `execute` method that changes and we would have two interfaces expressing the same concern for a very small gain.

Reply:
What is the opposition to having a separate `PreparedQuery` interface that extends the base `Query` interface? It seems like a good improvement to make the API more type-safe. We could catch these errors at compile-time as opposed to blowing up at runtime.

Reply:
that remains a valid point though, perhaps we could have `PreparedQuery` that extends `Query` and add `execute(Tuple)` and `executeBatch(List<Tuple>)`

Reply:
Agreed with @vietj here, and that would make sense to specialize `Query` here.

Reply:
yes, that's essentially what I was suggesting. `PreparedQuery` would extend `Query` and `PreparedQuery` would add the Tuple-overloaded execute methods. 

Sounds like we are on the same page now 👍